src/Database/LSMTree/Internal/RunReader.hs

```diff
@@ -64,18 +65,39 @@ data RunReader s fhandle = RunReader {
 new ::
      HasFS IO h
   -> HasBlockIO IO h
+  -> Maybe SerialisedKey  -- ^ offset
   -> Run.Run RealWorld (FS.Handle h)
   -> IO (RunReader RealWorld (FS.Handle h))
-new fs hbio readerRun = do
+new fs hbio mOffset readerRun = do
     readerKOpsHandle <-
       FS.hOpen fs (runKOpsPath (Run.runRunFsPaths readerRun)) FS.ReadMode
     -- double the file readahead window (only applies to this file descriptor)
     FS.hAdviseAll hbio readerKOpsHandle FS.AdviceSequential
-    readerCurrentEntryNo <- newPrimVar 0
-    -- load first page from disk, if it exists.
-    firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs readerKOpsHandle
-    readerCurrentPage <- newIORef firstPage
+
+    (page, entryNo) <- case mOffset of
+      Nothing -> do
+        -- load first page from disk, if it exists.
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs readerKOpsHandle
+        return (firstPage, 0)
+
+      Just offset -> do
+        let pageNo = Index.pageSpanStart (Index.search offset index)
+        -- load first page from disk, if it exists.
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+
+        case rawPageFindKey firstPage offset of
+          Just n ->
+            return (firstPage, n)
+          Nothing -> do
+            -- actually, it's the next page
+            nextPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+            return (nextPage, 0)
```

Comment:
We use `readDiskpage` in `next`, which means that we'll read the first page at an offset, but then start reading pages from the start. This is probably not caught by tests because we never start reading at an offset. We should extend the tests we already have to ensure we're testing this case properly

Reply:
Ah, yes, the handle still points at the beginning, because `hGetBufExactlyAt` does not change the handle state. Maybe we should use `hSeek` instead?

Alternatively, we could keep track of the page number ourselves and always use `hGetBufExactlyAt`.

Reply:
> you're going to have to either systematically use `readDiskPageAt` or use seeking

I don't think this has been resolved yet @recursion-ninja.

> Using `readDiskPageAt` and keeping track ourselves would get us closer to being able to use the same `O_DIRECT` file handle as the lookups use (we'd have to add our own manual read-ahead)

Do we want this? We save a file handle, but creating a new one for sequential reading and simply relying on the OS to do read-ahead seems quite convenient.

Reply:
This has been resolved by seeking, so we still rely on the handle's internal file pointer.